File: rulebook-cppcheck/rulebook_cppcheck/checkers/chain_call_wrap.py
from typing import override

from rulebook_cppcheck.checkers.rulebook_checkers import RulebookTokenChecker
from rulebook_cppcheck.messages import _Messages

try:
    from cppcheckdata import Token
except ImportError:
    from cppcheck.Cppcheck.addons.cppcheckdata import Token


class ChainCallWrapChecker(RulebookTokenChecker):
    """See detail: https://hanggrian.github.io/rulebook/rules/#chain-call-wrap"""
    ID: str = 'chain-call-wrap'
    _MSG_MISSING: str = 'chain.call.wrap.missing'
    _MSG_UNEXPECTED: str = 'chain.call.wrap.unexpected'

    _SHOULD_BREAK: set[str] = {';', '{', '}', '=', '?', ':'}
    _CLOSING_PARENTHESES: set[str] = {')', '}'}

    def __init__(self) -> None:
        super().__init__()
        self._reported_errors: set[tuple[str, int, int, str]] = set()

    @override
    def process_tokens(self, tokens: list[Token]) -> None:
        for token in [
            t for t in tokens if t.str == '.' and (not t.previous or t.previous.str != '.')
        ]:
            # check if this dot is inside function call parameters
            check_token: Token | None = token.previous
            paren_depth: int = 0
            while check_token:
                if check_token.str == ')':
                    paren_depth += 1
                elif check_token.str == '(':
                    if paren_depth == 0:
                        break
                    paren_depth -= 1
                elif check_token.str in self._SHOULD_BREAK:
                    check_token = None
                    break
                check_token = check_token.previous
            if check_token and check_token.str == '(':
                continue

            # look backwards to detect ternary
            check_token = token.previous
            paren_depth = 0
            found_ternary = False
            while check_token:
                if check_token.str == ')':
                    paren_depth += 1
                elif check_token.str == '(':
                    paren_depth -= 1
                elif paren_depth == 0:
                    if check_token.str == '?':
                        found_ternary = True
                        break
                    if check_token.str in self._SHOULD_BREAK:
                        break
                check_token = check_token.previous

            if found_ternary:
                continue

            # collect dots
            dots: list[Token] = []
            curr: Token | None = token
            depth: int = 0
            while curr:
                if curr.str == '(':
                    depth += 1
                elif curr.str == ')':
                    depth -= 1
                if depth < 0:
                    break
                if depth == 0:
                    if curr.str == '.':
                        dots.append(curr)
                    elif curr.str in self._SHOULD_BREAK:
                        break
                curr = curr.next

            # skip dots in single-line
            if len(dots) < 2:
                continue
            first_dot: Token = dots[0]
            if all(d.linenr == first_dot.linenr for d in dots):
                continue

            # checks for violation
            for dot in dots:
                prev_token: Token | None = dot.previous
                if not prev_token:
                    continue
                if prev_token.str in self._CLOSING_PARENTHESES and \
                    prev_token.previous and \
                    prev_token.linenr > prev_token.previous.linenr:
                    if dot.linenr > prev_token.linenr:
                        self._report_error_once(dot, _Messages.get(self._MSG_UNEXPECTED))
                    continue
                if dot.linenr == prev_token.linenr:
                    self._report_error_once(dot, _Messages.get(self._MSG_MISSING))

    def _report_error_once(self, token: Token, message: str) -> None:
        error_key: tuple[str, int, int, str] = (token.file, token.linenr, token.column, message)
        if error_key in self._reported_errors:
            return
        self._reported_errors.add(error_key)
        self.report_error(token, message)

File: rulebook-eslint/src/rules/chain-call-wrap.ts
import { AST, Rule } from 'eslint';
import { Expression, Super } from 'estree';
import RulebookRule from './rulebook-rules.js';
import messages from '../messages.js';
import Token = AST.Token;

/** {@link https://hanggrian.github.io/rulebook/rules/#chain-call-wrap|See detail} */
class ChainCallWrapRule extends RulebookRule {
    constructor() {
        super('chain-call-wrap', {
            [ChainCallWrapRule.MSG_MISSING]: messages.get(ChainCallWrapRule.MSG_MISSING),
            [ChainCallWrapRule.MSG_UNEXPECTED]: messages.get(ChainCallWrapRule.MSG_UNEXPECTED),
        });
    }

    create(context: Rule.RuleContext) {
        return {
            'CallExpression, MemberExpression'(node: Rule.Node) {
                // target root multiline chain call
                if (!node.parent ||
                    node.parent.type === 'CallExpression' ||
                    node.parent.type === 'MemberExpression') {
                    return;
                }
                if (node.loc!.start.line === node.loc!.end.line) {
                    return;
                }

                // collect dots
                const dots: Token[] = [];
                let current: Rule.Node | Expression | Super | null = node;
                while (current) {
                    let obj: Expression | Super | null = null;
                    let dot: Token | null = null;
                    if (current.type === 'CallExpression' &&
                        current.callee.type === 'MemberExpression') {
                        obj = current.callee.object;
                        dot = context.sourceCode.getTokenBefore(current.callee.property);
                    } else if (current.type === 'MemberExpression') {
                        obj = current.object;
                        dot = context.sourceCode.getTokenBefore(current.property);
                    }

                    if (dot && dot.value === '.') {
                        dots.push(dot);
                    }

                    current = obj;
                }

                // skip dots in single-line
                if (dots.length < 2) {
                    return;
                }
                const firstDot: Token = dots[0];
                if (dots.every(node => node.loc.start.line === firstDot.loc.start.line)) {
                    return;
                }

                // checks for violation
                dots.forEach(dot => {
                    const prevToken = context.sourceCode.getTokenBefore(dot);
                    if (!prevToken) {
                        return;
                    }
                    if ([')', '}', ']'].includes(prevToken.value) &&
                        prevToken.loc.start.line !==
                        context.sourceCode.getTokenBefore(prevToken)?.loc.end.line
                    ) {
                        if (dot.loc.start.line !== prevToken.loc.start.line) {
                            context.report({
                                node: dot,
                                messageId: ChainCallWrapRule.MSG_UNEXPECTED,
                            });
                        }
                    } else {
                        if (dot.loc.start.line !== prevToken.loc.end.line + 1) {
                            context.report({
                                node: dot,
                                messageId: ChainCallWrapRule.MSG_MISSING,
                            });
                        }
                    }
                });
            },
        };
    }

    static MSG_MISSING: string = 'chain.call.wrap.missing';
    static MSG_UNEXPECTED: string = 'chain.call.wrap.unexpected';
}

export default new ChainCallWrapRule();
